src/module.c

```diff
@@ -12244,6 +12244,134 @@ int RM_LoadConfigs(RedisModuleCtx *ctx) {
     return REDISMODULE_OK;
 }
 
+/* --------------------------------------------------------------------------
+ * ## RDB load/save API
+ * -------------------------------------------------------------------------- */
+
+#define REDISMODULE_RDB_STREAM_FILE 1
+
+typedef struct RedisModuleRdbStream {
+    int type;
+
+    union {
+        char *filename;
+    } data;
+} RedisModuleRdbStream;
+
+/* Create a stream object to save/load RDB to/from a file.
+ *
+ * This function returns a pointer to RedisModuleRdbStream which is owned
+ * by the caller. It requires a call to RM_RdbStreamFree() to free
+ * the object. */
+RedisModuleRdbStream *RM_RdbStreamCreateFromFile(const char *filename) {
+    RedisModuleRdbStream *stream = zmalloc(sizeof(*stream));
+    stream->type = REDISMODULE_RDB_STREAM_FILE;
+    stream->data.filename = zstrdup(filename);
+    return stream;
+}
+
+/* Release an RDB stream object. */
+void RM_RdbStreamFree(RedisModuleRdbStream *stream) {
+    switch (stream->type) {
+    case REDISMODULE_RDB_STREAM_FILE:
+        zfree(stream->data.filename);
+        break;
+    default:
+        serverAssert(0);
+        break;
+    }
+    zfree(stream);
+}
+
+/* Load RDB file from the `stream`. Dataset will be cleared first and then RDB
+ * file will be loaded.
+ *
+ * `flags` must be zero. This parameter is for future use.
+ *
+ * On success REDISMODULE_OK is returned, otherwise
+ * REDISMODULE_ERR is returned and errno is set to the following values:
+ *
+ * * EINVAL: `stream` is NULL or `flags` value is invalid.
+ * * EIO: Failed to load the RDB file. Check server logs for more info.
+ *
+ * Example:
+ *
+ *     RedisModuleRdbStream *s = RedisModule_RdbStreamCreateFromFile("exp.rdb");
+ *     RedisModule_RdbLoad(s, 0);
+ *     RedisModule_RdbStreamFree(s);
+ */
+int RM_RdbLoad(RedisModuleRdbStream *stream, int flags) {
+    if (!stream || flags != 0) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    }
+
+    if (server.aof_state != AOF_OFF) stopAppendOnly();
+
+    /* Kill existing RDB fork as it is saving outdated data. Also killing it
+     * will prevent COW memory issue. */
+    if (server.child_type == CHILD_TYPE_RDB) killRDBChild();
+
+    emptyData(-1,EMPTYDB_NO_FLAGS,NULL);
+
+    /* rdbLoad() can go back to the networking and process network events. If
+     * RM_RdbLoad() is called inside a command callback, we don't want to
+     * process the current client. Otherwise, we may free the client or try to
+     * process next message while we are already in the command callback. */
+    if (server.current_client) protectClient(server.current_client);
```

Comment:
we need to protect the calling client, not necessarily the executing one.
e.g. if a network client called a script which called a module command, we need to protect the network client.
not sure i understand the second part of your comment.

Reply:
I thought we normally used contexts for allowing future state to be added, as well as a hint to not call this command arbitrarily. Saving seems ok to call whenever, but loading it really should be in the context of the main thread.

Reply:
ohh, now i understand. you suggest to add a `RedisModuleContext` argument to the API, even if we don't currently need it, and this way if we'll need something from it in the future, we'll have it?
if we'll want to know the redis server state, we don't need that, but if we'd like to know anything about the calling module, we will...

looking at other APIs without a context: RM_GetClientInfoById, RM_AvoidReplicaTraffic, RM_Alloc, RM_LatencyAddSample, RM_Milliseconds, RM_GetMyClusterID, RM_SendChildHeartbeat, RM_EventLoopAdd.

i see many of them are about getting info, and not about performing actions, and the ones that are about performing actions, these are very raw actions, so i suppose it's a good idea to require a context.
p.s. i remember one case that we used to ask a context and later realized it was an issue: RM_Log.

Reply:
since the RedisModuleRdbStream is later passed to the RM_RdbLoad API, we can add the context just there, and then we can extract the info we want from it. e.g. similar to what we do with RedisModuleIO.

however, i personally think it's unnecessary in RM_RdbStreamCreateFromFile, and only needed in the actual save and load API.
@madolson @yossigo WDYT?

Reply:
I think just the Load/Save APIs is really what is needed. The other APIs are pretty benign.

Reply:
added ctx to load/save functions.